File: legacy/components/desas/app/analyzer/forensics.py
import math
import re
import zipfile
import io
import logging
from collections import Counter
from typing import Dict, List, Any

logger = logging.getLogger("uvicorn")

# -------- CORE ANALYTICS --------

def shannon_entropy(data: bytes) -> float:
    """Calculates the Shannon entropy of a byte string."""
    if not data:
        return 0.0
    freq = Counter(data)
    probs = [c / len(data) for c in freq.values()]
    return -sum(p * math.log2(p) for p in probs)

def obfuscation_heuristics(data: str) -> int:
    """Detects common obfuscation patterns in scripts (VBA, JS, etc.)"""
    score = 0
    if not data:
        return 0
        
    low_data = data.lower()
    
    # Character code usage
    if re.search(r'chr\(\d+\)', low_data):
        score += 20
        
    # Heavy string concatenation
    if data.count("+") > 50 or data.count("&") > 50:
        score += 15
        
    # Long encoded-looking blobs (Base64-ish)
    if re.search(r'[A-Za-z0-9+/=]{80,}', data):
        score += 25
        
    # Mixed case/randomized case detection (heuristics)
    # If the ratio of uppercase to lowercase is very balanced in a large string, it might be obfuscated
    if len(data) > 100:
        alphas = [c for c in data if c.isalpha()]
        if alphas:
            u_count = sum(1 for c in alphas if c.isupper())
            ratio = u_count / len(alphas)
            if 0.3 < ratio < 0.7:
                score += 10
                
    return score

def vba_semantic_analysis(vba_code: str) -> Dict[str, Any]:
    """Analyzes VBA code intent beyond simple existence."""
    indicators = {
        "auto_exec": False,
        "env_fingerprinting": False,
        "staging_logic": False,
        "obfuscation": False,
        "suspicious_calls": []
    }
    
    if not vba_code:
        return indicators
        
    vba_low = vba_code.lower()
    
    # 1. Execution Triggers
    auto_execs = ["auto_open", "workbook_open", "document_open", "auto_exec", "autoopen"]
    indicators["auto_exec"] = any(x in vba_low for x in auto_execs)
    
    # 2. Environment Awareness (Anti-VM / Anti-Sandbox)
    env_calls = ["environ", "username", "computername", "getobject(\"winmgmts:", "win32_process"]
    indicators["env_fingerprinting"] = any(x in vba_low for x in env_calls)
    
    # 3. Payload Staging
    staging_keywords = ["split(", "join(", "replace(", "chr(", "strconv(", "getbyte"]
    # If multiple staging keywords are used heavily
    indicators["staging_logic"] = sum(1 for x in staging_keywords if x in vba_low) >= 2
    
    # 4. Critical External Calls
    susp_calls_map = {
        "shell": "Shell execution",
        "wscript.shell": "WScript interaction",
        "powershell": "PowerShell invocation",
        "createobject": "Dynamic object creation",
        "adodb.stream": "File system stream (Dropper logic)",
        "xmlhttp": "Network download (XMLHTTP)",
        "winhttprequest": "Network download (WinHTTP)",
        "urlretrieve": "Network download",
        "base64": "Encoded payload"
    }
    
    for call, desc in susp_calls_map.items():
        if call in vba_low:
            indicators["suspicious_calls"].append(desc)
            
    # 5. Obfuscation check
    indicators["obfuscation"] = vba_low.count("chr(") > 10 or "strreverse" in vba_low
    
    return indicators

def structural_analysis(content: bytes, filename: str) -> Dict[str, Any]:
    """Analyzes the file structure for anomalies (ZIP/Office)."""
    findings = {
        "extra_data": False, # ZIP overlay/anomaly
        "unexpected_files": [],
        "structure_score": 0,
        "is_archive": False
    }
    
    if not content:
        return findings
        
    # Check if it's a ZIP-based file
    if zipfile.is_zipfile(io.BytesIO(content)):
        findings["is_archive"] = True
        try:
            with zipfile.ZipFile(io.BytesIO(content)) as z:
                for f in z.infolist():
                    # Check for potential payload staging in archives
                    ext = "." + f.filename.split(".")[-1].lower() if "." in f.filename else ""
                    risky_exts = [".exe", ".dll", ".js", ".vbs", ".ps1", ".vbe", ".jse", ".cmd", ".bat"]
                    
                    if ext in risky_exts:
                        findings["unexpected_files"].append(f.filename)
                        findings["structure_score"] += 30
                        
                    # Double extension check
                    if re.search(r'\.[a-z0-9]{2,4}\.(exe|js|vbs|bat|cmd)$', f.filename.lower()):
                        findings["unexpected_files"].append(f"{f.filename} (Potential Masquerading)")
                        findings["structure_score"] += 20
                        
                    # Encrypted/Password protected files
                    if f.flag_bits & 0x1:
                        findings["unexpected_files"].append(f"{f.filename} (Encrypted)")
                        findings["structure_score"] += 15
                        
        except Exception as e:
            logger.warning(f"Structural analysis failed for {filename}: {e}")
            
    return findings

def detect_polyglots(content: bytes) -> List[str]:
    """Identifies multiple file signatures in a single file buffer."""
    findings = []
    if len(content) < 100:
        return findings
        
    signatures = {
        b"PK\x03\x04": "ZIP/Office",
        b"%PDF": "PDF",
        b"\xff\xd8\xff": "JPEG",
        b"\x89PNG\r\n\x1a\n": "PNG",
        b"MZ": "PE Executable",
        b"BM": "BMP",
        b"GIF8": "GIF",
        b"\x7fELF": "ELF Executable"
    }
    
    found = []
    for sig, label in signatures.items():
        if sig in content:
            found.append(label)
            
    if len(found) > 1:
        # Check if they are sufficiently far apart to not be false positives or standard nesting
        # A PDF can contain a ZIP (ObjStm), but a PDF starting with MZ at offset 0 is a polyglot
        findings.append(f"Polyglot Indicators: {', '.join(found)}")
        
    # Check for Appended Data (EOF Anomaly)
    # If a PNG has a PE/ZIP signature at the very end
    image_sigs = [b"\xff\xd9", b"\x00\x00\x00\x00IEND\xaeB`\x82"] # JPEG/PNG EOF
    for eof in image_sigs:
        if eof in content:
            eof_pos = content.rfind(eof) + len(eof)
            if len(content) > eof_pos + 10:
                # Extra data found after image EOF
                extra = content[eof_pos:]
                if any(s in extra for s in [b"MZ", b"PK\x03\x04", b"powershell"]):
                    findings.append("Appended malicious payload detected after image EOF")
                    
    return findings

def analyze_ole_streams(content: bytes) -> Dict[str, Any]:
    """Scans OLE streams for specific malicious indicators beyond basic macros."""
    import olefile
    indicators = {
        "risky_streams": [],
        "ole_anomalies": False
    }
    
    if not olefile.isOleFile(io.BytesIO(content)):
        return indicators
        
    try:
        ole = olefile.OleFileIO(io.BytesIO(content))
        streams = [s[0] if isinstance(s, list) else s for s in ole.listdir()]
        
        target_streams = {
            "MBDG": "Potentially obfuscated shellcode stream",
            "Package": "Embedded OLE Package (often an EXE)",
            "Equation Native": "Equation Editor Exploit (CVE-2017-11882)",
            "Ole10Native": "Embedded Native code/binary",
            "ObjectPool": "Nested OLE objects (Evasion)"
        }
        
        for name, desc in target_streams.items():
            if any(name.lower() in str(s).lower() for s in streams):
                indicators["risky_streams"].append(desc)
                indicators["ole_anomalies"] = True
                
    except Exception as e:
        logger.warning(f"OLE Stream analysis failed: {e}")
        
    return indicators

def detect_xlm_macros(vba_code: str, raw_bytes: bytes) -> List[str]:
    """Detects Legacy Excel 4.0 (XLM) macros which are often used for evasion."""
    findings = []
    # XLM macros often use specific functions that show up in raw bytes or strings
    xlm_patterns = [
        rb"EXEC", rb"REGISTER", rb"CALL", rb"HALT", rb"RUN", rb"CHAR",
        rb"GET.WINDOW", rb"WINDOW.HIDE", rb"WORKBOOK.HIDE"
    ]
    
    found_ops = []
    for p in xlm_patterns:
        if p in raw_bytes:
            found_ops.append(p.decode())
            
    if len(found_ops) >= 3:
        findings.append(f"Excel 4.0 (XLM) Macro pattern detected: {', '.join(found_ops)}")
        
    return findings

def pdf_forensic_signals(text_content: str, raw_bytes: bytes) -> Dict[str, Any]:
    """Deep PDF analysis for evasion and suspicious objects."""
    signals = {
        "has_js": False,
        "has_launch": False,
        "embedded_files": False,
        "evasion_detected": False,
        "suspicious_tags": []
    }
    
    # We use regex on raw bytes for tags that usually trigger SOC hits
    try:
        raw_str = raw_bytes.decode('utf-8', errors='ignore').lower()
        
        tag_map = {
            "/javascript": ("has_js", "Javascript code"),
            "/js": ("has_js", "Javascript code (short)"),
            "/launch": ("has_launch", "Launch action (EXE execution)"),
            "/embeddedfile": ("embedded_files", "Embedded file payload"),
            "/richtext": ("suspicious_tags", "Rich text content"),
            "/openaction": ("suspicious_tags", "Auto-open action"),
            "/aa": ("suspicious_tags", "Additional Action (Auto-trigger)")
        }
        
        for tag, (key, desc) in tag_map.items():
            if tag in raw_str:
                if isinstance(signals[key], bool):
                    signals[key] = True
                else:
                    signals[key].append(desc)
                    
        # Evasion detection (e.g., ObjStm without XRef - very basic check)
        if "/objstm" in raw_str and "/xref" not in raw_str:
            signals["evasion_detected"] = True
            
    except Exception:
        pass
        
    return signals

def infer_mitre_techniques(signals: Dict[str, Any]) -> List[str]:
    """Maps forensic signals to MITRE ATT\u0026CK techniques."""
    tech_map = {
        "auto_exec": "T1204.002", # User Execution: Malicious File
        "obfuscation": "T1027",    # Obfuscated Files or Information
        "env_fingerprinting": "T1497", # Virtualization/Sandbox Evasion
        "staging_logic": "T1059",  # Command and Scripting Interpreter
        "has_macros": "T1059.005", # Visual Basic
        "has_js": "T1059.007",     # JavaScript
        "double_extension": "T1036.007", # Masquerading: Double Extension
        "zip_then_excel": "T1566.001", # Phishing: Spearphishing Attachment
        "polyglot": "T1027.001",   # Binary Padding / Polyglot
        "ole_anomaly": "T1204.002", # Malicious File (OLE)
        "xlm_macro": "T1059",      # Command and Scripting Interpreter
        "appended_data": "T1027.001" # Binary Padding
    }
    
    techniques = []
    for sig, tech in tech_map.items():
        if signals.get(sig):
            techniques.append(tech)
            
    return sorted(list(set(techniques)))

def analyze_image_forensics(exif_data: Dict[str, Any], raw_bytes: bytes) -> Dict[str, Any]:
    """Heuristic check for steganography or suspicious metadata."""
    results = {
        "suspicious": False,
        "reasons": []
    }
    
    # Check entropy of the entire image (high entropy can indicate hidden payloads)
    e = shannon_entropy(raw_bytes)
    if e > 7.9:
        results["suspicious"] = True
        results["reasons"].append("Extremely high entropy (Potential Payload)")
        
    # Check EXIF for scripty looking strings
    if exif_data:
        script_indicators = ["script", "powershell", "curl", "wget", "http", "<?php", "exec("]
        for val in exif_data.values():
            if isinstance(val, str):
                v_low = val.lower()
                if any(x in v_low for x in script_indicators):
                    results["suspicious"] = True
                    results["reasons"].append(f"Suspicious string in metadata: {val[:30]}...")
                    
    return results

File: src/lib/forensics/forensics.ts
import * as CFB from "cfb";
import type { CFB$Container } from "cfb";
import JSZip from "jszip";

// Ported from desas's analyzer/forensics.py — pure byte/string forensic
// heuristics, portable without any native dependency once the
// OLE-container listing (cfb, replacing Python's olefile) and ZIP
// listing (jszip, replacing Python's built-in zipfile) are swapped in.

export function shannonEntropy(data: Buffer): number {
  if (data.length === 0) return 0;
  const freq = new Map<number, number>();
  for (const byte of data) freq.set(byte, (freq.get(byte) ?? 0) + 1);
  let entropy = 0;
  for (const count of freq.values()) {
    const p = count / data.length;
    entropy -= p * Math.log2(p);
  }
  return entropy;
}

export function obfuscationHeuristics(data: string): number {
  if (!data) return 0;
  let score = 0;
  const low = data.toLowerCase();

  if (/chr\(\d+\)/.test(low)) score += 20;
  if ((data.match(/\+/g)?.length ?? 0) > 50 || (data.match(/&/g)?.length ?? 0) > 50) score += 15;
  if (/[A-Za-z0-9+/=]{80,}/.test(data)) score += 25;

  if (data.length > 100) {
    const alphas = [...data].filter((c) => /[a-zA-Z]/.test(c));
    if (alphas.length > 0) {
      const upperCount = alphas.filter((c) => c === c.toUpperCase() && c !== c.toLowerCase()).length;
      const ratio = upperCount / alphas.length;
      if (ratio > 0.3 && ratio < 0.7) score += 10;
    }
  }

  return score;
}

export interface VbaSemanticResult {
  autoExec: boolean;
  envFingerprinting: boolean;
  stagingLogic: boolean;
  obfuscation: boolean;
  suspiciousCalls: string[];
}

const AUTO_EXEC_MARKERS = ["auto_open", "workbook_open", "document_open", "auto_exec", "autoopen"];
const ENV_CALL_MARKERS = ["environ", "username", "computername", 'getobject("winmgmts:', "win32_process"];
const STAGING_MARKERS = ["split(", "join(", "replace(", "chr(", "strconv(", "getbyte"];
const SUSPICIOUS_CALL_MAP: Record<string, string> = {
  shell: "Shell execution",
  "wscript.shell": "WScript interaction",
  powershell: "PowerShell invocation",
  createobject: "Dynamic object creation",
  "adodb.stream": "File system stream (dropper logic)",
  xmlhttp: "Network download (XMLHTTP)",
  winhttprequest: "Network download (WinHTTP)",
  urlretrieve: "Network download",
  base64: "Encoded payload",
};

export function vbaSemanticAnalysis(vbaCode: string): VbaSemanticResult {
  const result: VbaSemanticResult = {
    autoExec: false,
    envFingerprinting: false,
    stagingLogic: false,
    obfuscation: false,
    suspiciousCalls: [],
  };
  if (!vbaCode) return result;

  const low = vbaCode.toLowerCase();
  result.autoExec = AUTO_EXEC_MARKERS.some((m) => low.includes(m));
  result.envFingerprinting = ENV_CALL_MARKERS.some((m) => low.includes(m));
  result.stagingLogic = STAGING_MARKERS.filter((m) => low.includes(m)).length >= 2;

  for (const [call, desc] of Object.entries(SUSPICIOUS_CALL_MAP)) {
    if (low.includes(call)) result.suspiciousCalls.push(desc);
  }

  result.obfuscation = (low.match(/chr\(/g)?.length ?? 0) > 10 || low.includes("strreverse");
  return result;
}

export interface StructuralAnalysis {
  isArchive: boolean;
  unexpectedFiles: string[];
  structureScore: number;
}

const RISKY_EXTENSIONS = [".exe", ".dll", ".js", ".vbs", ".ps1", ".vbe", ".jse", ".cmd", ".bat"];

export async function structuralAnalysis(content: Buffer): Promise<StructuralAnalysis> {
  const findings: StructuralAnalysis = { isArchive: false, unexpectedFiles: [], structureScore: 0 };

  let zip: JSZip;
  try {
    zip = await JSZip.loadAsync(content);
  } catch {
    return findings;
  }

  findings.isArchive = true;
  for (const filename of Object.keys(zip.files)) {
    const lower = filename.toLowerCase();
    const ext = lower.includes(".") ? `.${lower.split(".").pop()}` : "";

    if (RISKY_EXTENSIONS.includes(ext)) {
      findings.unexpectedFiles.push(filename);
      findings.structureScore += 30;
    }
    if (/\.[a-z0-9]{2,4}\.(exe|js|vbs|bat|cmd)$/.test(lower)) {
      findings.unexpectedFiles.push(`${filename} (potential masquerading)`);
      findings.structureScore += 20;
    }
  }

  return findings;
}

const FILE_SIGNATURES: [Buffer, string][] = [
  [Buffer.from([0x50, 0x4b, 0x03, 0x04]), "ZIP/Office"],
  [Buffer.from("%PDF"), "PDF"],
  [Buffer.from([0xff, 0xd8, 0xff]), "JPEG"],
  [Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]), "PNG"],
  [Buffer.from("MZ"), "PE Executable"],
  [Buffer.from("BM"), "BMP"],
  [Buffer.from("GIF8"), "GIF"],
  [Buffer.from([0x7f, 0x45, 0x4c, 0x46]), "ELF Executable"],
];

export function detectPolyglots(content: Buffer): string[] {
  const findings: string[] = [];
  if (content.length < 100) return findings;

  const found = FILE_SIGNATURES.filter(([sig]) => content.includes(sig)).map(([, label]) => label);
  if (found.length > 1) {
    findings.push(`Polyglot indicators: ${found.join(", ")}`);
  }

  const jpegEof = Buffer.from([0xff, 0xd9]);
  const pngEof = Buffer.from("\x00\x00\x00\x00IEND\xaeB\x60\x82", "binary");
  for (const eof of [jpegEof, pngEof]) {
    const eofPos = content.lastIndexOf(eof);
    if (eofPos === -1) continue;
    const dataEnd = eofPos + eof.length;
    if (content.length > dataEnd + 10) {
      const extra = content.subarray(dataEnd);
      if (
        extra.includes(Buffer.from("MZ")) ||
        extra.includes(Buffer.from([0x50, 0x4b, 0x03, 0x04])) ||
        extra.includes(Buffer.from("powershell"))
      ) {
        findings.push("Appended malicious payload detected after image EOF");
      }
    }
  }

  return findings;
}

export interface OleStreamResult {
  riskyStreams: string[];
  oleAnomalies: boolean;
}

const OLE_TARGET_STREAMS: Record<string, string> = {
  mbdg: "Potentially obfuscated shellcode stream",
  package: "Embedded OLE Package (often an EXE)",
  "equation native": "Equation Editor exploit (CVE-2017-11882)",
  ole10native: "Embedded native code/binary",
  objectpool: "Nested OLE objects (evasion)",
};

export function analyzeOleStreams(content: Buffer): OleStreamResult {
  const indicators: OleStreamResult = { riskyStreams: [], oleAnomalies: false };

  let container: CFB$Container;
  try {
    container = CFB.read(content, { type: "buffer" });
  } catch {
    return indicators;
  }

  const streamNames = container.FileIndex.map((e) => e.name.toLowerCase());
  for (const [marker, desc] of Object.entries(OLE_TARGET_STREAMS)) {
    if (streamNames.some((s) => s.includes(marker))) {
      indicators.riskyStreams.push(desc);
      indicators.oleAnomalies = true;
    }
  }

  return indicators;
}

const XLM_PATTERNS = ["EXEC", "REGISTER", "CALL", "HALT", "RUN", "CHAR", "GET.WINDOW", "WINDOW.HIDE", "WORKBOOK.HIDE"];

export function detectXlmMacros(rawContent: Buffer): string[] {
  const asString = rawContent.toString("latin1");
  const found = XLM_PATTERNS.filter((p) => asString.includes(p));
  if (found.length >= 3) {
    return [`Excel 4.0 (XLM) macro pattern detected: ${found.join(", ")}`];
  }
  return [];
}

export interface PdfForensicSignals {
  hasJs: boolean;
  hasLaunch: boolean;
  embeddedFiles: boolean;
  evasionDetected: boolean;
  suspiciousTags: string[];
}

export function pdfForensicSignals(rawContent: Buffer): PdfForensicSignals {
  const signals: PdfForensicSignals = {
    hasJs: false,
    hasLaunch: false,
    embeddedFiles: false,
    evasionDetected: false,
    suspiciousTags: [],
  };

  const raw = rawContent.toString("latin1").toLowerCase();

  if (raw.includes("/javascript") || raw.includes("/js")) signals.hasJs = true;
  if (raw.includes("/launch")) signals.hasLaunch = true;
  if (raw.includes("/embeddedfile")) signals.embeddedFiles = true;
  if (raw.includes("/richtext")) signals.suspiciousTags.push("Rich text content");
  if (raw.includes("/openaction")) signals.suspiciousTags.push("Auto-open action");
  if (raw.includes("/aa")) signals.suspiciousTags.push("Additional action (auto-trigger)");
  if (raw.includes("/objstm") && !raw.includes("/xref")) signals.evasionDetected = true;

  return signals;
}

const MITRE_TECHNIQUE_MAP: Record<string, string> = {
  autoExec: "T1204.002",
  obfuscation: "T1027",
  envFingerprinting: "T1497",
  stagingLogic: "T1059",
  hasMacros: "T1059.005",
  hasJs: "T1059.007",
  doubleExtension: "T1036.007",
  polyglot: "T1027.001",
  oleAnomaly: "T1204.002",
  xlmMacro: "T1059",
  appendedData: "T1027.001",
};

export function inferMitreTechniques(signals: Record<string, boolean>): string[] {
  const techniques = Object.entries(MITRE_TECHNIQUE_MAP)
    .filter(([sig]) => signals[sig])
    .map(([, tech]) => tech);
  return [...new Set(techniques)].sort();
}
